### PR TITLE
Ticket5122 scriptgen multiple lines

### PR DESCRIPTION
**Pull request by adalator**
### Description of work

Allows selecting multiple rows in the script generator using CTRL/SHIFT.

### Ticket

https://github.com/ISISComputingGroup/IBEX/issues/5122

### Acceptance criteria

* Multiple rows can be selected and deleted
* Multiple rows can be selected and duplicated. 
* Multiple rows can be selected and moved up or down.
* Unit tests run

### Unit tests

JUnit tests added in ibex/scriptgenerator/tests/ActionsTableTest.java

### System tests

See https://github.com/ISISComputingGroup/System_Tests_UI_E4/pull/61

### Documentation
None

---

#### Code Review

- [ ] Is the code of an acceptable quality?
- [ ] Do the changes function as described and is it robust?
- [ ] Have the changes been documented in the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev)?

### Final Steps
- [ ] Reviewer has moved the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev) entry for this ticket in the "Changes merged into master" section

### CODE REVIEW

File: base/uk.ac.stfc.isis.ibex.ui.scriptgenerator/src/uk/ac/stfc/isis/ibex/ui/scriptgenerator/views/ActionsViewTable.java [hunk 1]
```diff
@@ -169,7 +169,7 @@ public void setRows(Collection<ScriptGeneratorAction> rows) {
 		if ((!viewer.getTable().isDisposed())) {
```
Comment (adalator) [inline base/uk.ac.stfc.isis.ibex.ui.scriptgenerator/src/uk/ac/stfc/isis/ibex/ui/scriptgenerator/views/ActionsViewTable.java:169]: Commenting out this section will restore the original behaviour of the code and give you an idea of what it should look like.
```suggestion
/*
@Override
	public void setRows(Collection<ScriptGeneratorAction> rows) {
		if ((!viewer.getTable().isDisposed())) {
			int focusRow = viewer.getTable().getSelectionIndex();
			int focusColumn = 0;

			
			if (shiftCellFocusToNewlyAddedRow) {
				focusRow = viewer.getTable().getSelectionIndex() + 1;
				shiftCellFocusToNewlyAddedRow = false;
			// if row is not empty
			} else if (focusRow != -1) {
				focusColumn = viewer.getColumnViewerEditor().getFocusCell().getColumnIndex();
			}
			
			viewer.setInput(new WritableList<ScriptGeneratorAction>(rows, null));
			setCellFocus(focusRow, focusColumn);
		}
	} 
	*/
	/**
	 * Sets focus of cell
	 * @param row row number of table
	 * @param column column number of table
	 */
	 /*
	private void setCellFocus(int row, int column) {
		if (row >= 0) {
			viewer.editElement(viewer.getElementAt(row), column);
		}
	}
        */
```

File: base/uk.ac.stfc.isis.ibex.ui.scriptgenerator/src/uk/ac/stfc/isis/ibex/ui/scriptgenerator/views/ScriptGeneratorView.java [hunk 2]
```diff
@@ -288,13 +288,13 @@ private void displayLoaded() {
 			    // Make buttons to move an action up and down the list
 		        Button btnMoveActionUp = createMoveRowButton(moveComposite, "move_up.png", "up");
 		        btnMoveActionUp.addListener(SWT.Selection, e -> {
-		        	scriptGeneratorViewModel.moveActionUp(table.getSelectionIndex());
+		        	scriptGeneratorViewModel.moveActionUp(table.getSelectionIndices());
 		        	table.setSelectionIndex(Math.max(0, table.getSelectionIndex() - 1));
```
Comment (adalator) [inline base/uk.ac.stfc.isis.ibex.ui.scriptgenerator/src/uk/ac/stfc/isis/ibex/ui/scriptgenerator/views/ScriptGeneratorView.java:292]: There should be some code here that gets the original selection, moves it up one providing the selection isn't already at the top (at index 0), and then sets that as the new selection once `moveActionUp()` has been called. The same should be down for the move button down listener below.

File: base/uk.ac.stfc.isis.ibex.ui.scriptgenerator/src/uk/ac/stfc/isis/ibex/ui/scriptgenerator/views/ScriptGeneratorView.java [hunk 3]
```diff
@@ -289,16 +292,17 @@ private void displayLoaded() {
...
 		        });
 		        
 		        Button btnMoveActionDown = createMoveRowButton(moveComposite, "move_down.png", "down");
 		        btnMoveActionDown.addListener(SWT.Selection, e -> {
-		        	scriptGeneratorViewModel.moveActionDown(table.getSelectionIndex());
-		            table.setSelectionIndex(Math.min(
-		            		table.getSelectionIndex() + 1, scriptGeneratorViewModel.getActions().size()));
+		        	List<ScriptGeneratorAction> selection = table.selectedRows();
+		        	scriptGeneratorViewModel.moveActionDown(selection);
+		        	table.viewer().setSelection((ISelection) selection);
```
Comment (Tom-Willemsen) [inline base/uk.ac.stfc.isis.ibex.ui.scriptgenerator/src/uk/ac/stfc/isis/ibex/ui/scriptgenerator/views/ScriptGeneratorView.java:305]: I'm getting the following exception on this line when selecting multiple rows and moving them up/down.

```
!ENTRY org.eclipse.ui 4 0 2020-06-19 14:34:15.174
!MESSAGE Unhandled event loop exception
!STACK 0
java.lang.ClassCastException: class java.util.ArrayList cannot be cast to class org.eclipse.jface.viewers.ISelection (java.util.ArrayList is in module java.base of loader 'bootstrap'; org.eclipse.jface.viewers.ISelection is in unnamed module of loader org.eclipse.osgi.internal.loader.EquinoxClassLoader @41a90fa8)
	at uk.ac.stfc.isis.ibex.ui.scriptgenerator.views.ScriptGeneratorView.lambda$9(ScriptGeneratorView.java:305)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4175)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1057)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:3988)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3587)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1160)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1049)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:660)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:559)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:154)
	at uk.ac.stfc.isis.ibex.scriptgenerator.product.Application.start(Application.java:47)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:137)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:107)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:401)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:657)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:594)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1465)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1438)
```

File: base/uk.ac.stfc.isis.ibex.scriptgenerator/src/uk/ac/stfc/isis/ibex/scriptgenerator/table/ActionsTable.java [hunk 4]
```diff
@@ -162,30 +160,45 @@ public void clearAction() {
...
-			
-			if (newIndex < 0) {
-				newIndex = 0;
-			} else if (newIndex >= this.actions.size()) {
-				newIndex = this.actions.size() - 1;
-			}
-			
-			Collections.swap(newActions, oldIndex, newIndex);
+	public void moveActionUp(List<ScriptGeneratorAction> actionsToMove) {
+		if (actionsToMove.contains(actions.get(0))) {
```
Comment (Tom-Willemsen) [inline base/uk.ac.stfc.isis.ibex.scriptgenerator/src/uk/ac/stfc/isis/ibex/scriptgenerator/table/ActionsTable.java:191]: I get the following error if I have no actions in the table and press the move up/mode down buttons. I think we should fix this by greying out the buttons if there are no actions in the table/no actions selected.

```
*2020-06-19 14:37:55.819 [main] ERROR org.eclipse.ui.workbench - Index 0 out of bounds for length 0
	java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:64)
	java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:70)
	java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:248)
	java.base/java.util.Objects.checkIndex(Objects.java:372)
	java.base/java.util.ArrayList.get(ArrayList.java:458)
	uk.ac.stfc.isis.ibex.scriptgenerator.table.ActionsTable.moveActionUp(ActionsTable.java:191)
	uk.ac.stfc.isis.ibex.scriptgenerator.ScriptGeneratorSingleton.moveActionUp(ScriptGeneratorSingleton.java:473)
	uk.ac.stfc.isis.ibex.ui.scriptgenerator.views.ScriptGeneratorViewModel.moveActionUp(ScriptGeneratorViewModel.java:323)
	uk.ac.stfc.isis.ibex.ui.scriptgenerator.views.ScriptGeneratorView.lambda$8(ScriptGeneratorView.java:297)
	org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	org.eclipse.swt.widgets.Display.sendEvent(Display.java:4175)
	org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1057)
	org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:3988)
	org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3587)
	org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1160)
	org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1049)
	org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)
	org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:660)
	org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338)
	org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:559)
	org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:154)
	uk.ac.stfc.isis.ibex.scriptgenerator.product.Application.start(Application.java:47)
	org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203)
	org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:137)
	org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:107)
	org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:401)
	org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
	java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	java.base/java.lang.reflect.Method.invoke(Method.java:566)
	org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:657)
	org.eclipse.equinox.launcher.Main.basicRun(Main.java:594)
	org.eclipse.equinox.launcher.Main.run(Main.java:1465)
	org.eclipse.equinox.launcher.Main.main(Main.java:1438)

```

**Comment by Tom-Willemsen** [comment]
I added release notes for you.

**Pull request merged**